File: PYTHON/5-BinarySearch/1-binary_search.py
from typing import List
import bisect

def search(nums: List[int], target: int) -> int:
    lista = enumerate(nums)
    for i, num in lista:
        if num == target:
            return i
    return -1

# # melhor
# def search(nums: List[int], target: int) -> int:
#     left, right = 0, len(nums) - 1
#     while left <= right:
#         mid = left + ((right - left) // 2)
#         if nums[mid] == target:
#             return mid
#         elif nums[mid] < target:
#             left = mid + 1
#         else:
#             right = mid - 1
#     return -1

# # usando bisect
# def search(nums: List[int], target: int) -> int:
#     index = bisect.bisect_left(nums, target)
#     return index if index < len(nums) and nums[index] == target else -1

if __name__ == "__main__":
    nums = [-1, 0, 2, 4, 6, 8]
    target = 4
    print(search(nums, target))

File: CPP/5-BinarySearch/1-binary_search.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// brute force
int search(vector<int>& nums, int target) {
    int left = 0;
    int right = nums.size() - 1;

    while (left <= right) {
        int mid = (left + right) / 2;
        if (nums[mid] == target) {
            return mid;
        } else if (nums[mid] < target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return -1;
}

int main() {
    vector<int> nums = {-1, 0, 2, 4, 6, 8};
    int target = 4;
    cout << search(nums, target);
}